src/global/ValueId.h

```diff
@@ -326,10 +326,10 @@ class ValueId {
   /// This operator is only for debugging and testing. It returns a
   /// human-readable representation.
   friend std::ostream& operator<<(std::ostream& ostr, const ValueId& id) {
-    ostr << toString(id.getDatatype()) << ':';
+    ostr << toString(id.getDatatype())[0] << ':';
     auto visitor = [&ostr]<typename T>(T&& value) {
       if constexpr (ad_utility::isSimilar<T, ValueId::UndefinedType>) {
-        ostr << "Undefined";
+        ostr << "xx";
```

Comment:
The better way would be to first do
```
if (isUndefined) { ostr << id.getBits()}
```
and then just assert in the visitor, that the undef was handleed before (AD_FAIL).
S.T. the normal undefined becomes U:0  , we currently use other values with the undefined types for special internal IDs, those are currently undebugabble.

Reply:
I am not sure that I understand what you mean, so let me paraphrase what I understood. There is `Undefined` type for `ValudId` with the value `0000`. This type normally only has one value - (`UNDEF`?) with all zero bits. Currently the `Undefined` type is also used with non zero data bits. You want to get the raw bits for all other undefineds.

The visitor only looks at the data type bits and would also get undefined with non zero data bits. `isUndefined` on the other hand only matches the undefined with all zero data bits.

Reply:
As I just saw, you already did it the right way.
So I will just wait forthe checks to run throug again, and then we can merge this.

Reply:
What are your thoughts on adding an additional test that tests exactly this behavior for undefined with non zero data? The behavior seems to be used in practice although it is hacky.

Reply:
Yes of course:)
I am all for tests, especially such non-obvious ones.